## Используемые библиотеки

In [74]:
import numpy as np
import math
import pandas as pd

## Нужные функции

In [75]:
# округление 
def o(x):
    return round(x * pow(10,5))/pow(10,5) 

# значение функции 
def getValue(x):
    return math.exp(2*x)

# точное значение певрой производной
def getExpFirstDerivative (x):
    return 2 * math.exp(2*x)

def getExpSecondDerivative (x):
    return 4 *  math.exp(2*x)


# значение первой производной с первый порядком точности 
def getFirstDevivativeOH(x, h):
    return (o(getValue(x)) - o(getValue(x-h)))/h

# значение первой производной со вторым порядком точности 
def getFirstDerivativeOH2(x, h):
    return ((-3)* o(getValue(x)) + 4* o(getValue(x + h)) - o(getValue(x + 2*h)))/(2*h)

# значение второй производной с первым порядком точности 
def getSecondDerivativeOH2(x,  h):
    return (o(getValue(x+h)) - 2*o(getValue(x)) + o(getValue(x-h)))/(pow(h, 2))

# значение первой производнйо с третьим порядком точности 
def getFirstDerivativeOH3(x,  h):
    return ((-11)*o(getValue(x-3*h)) +  18*o(getValue(x-2*h)) - 9*o(getValue(x-h)) + 2*o(getValue(x)))/(6*h)



## Задания

<img src='task.PNG'>


### 1) Выполнить пункт 1 из задания

<img src='task2.PNG'>

In [76]:
df_1 = pd.DataFrame(columns = ["x", 
                             "f(x)",
                             "f(x)'",
                              "f_'(x)",
                              "O(h)",
                              "f='(x)",
                              "O(h^2)",
                              "f(x)''",
                              "f_(x)''",
                              "O(h^2)"]) 

In [77]:
def f_1(df):
    h = 0.0 # шаг
    n = 0 # количтсво узлов
    begin = 0.0 # начало отрезка 
    h = 0.1
    n = 10
    
    for i in range(n):
        x = begin + h * (i)
        derivf = getExpFirstDerivative(x)
        secondderivf = getExpSecondDerivative(x)
        df = df.append({"x" : x, 
                        "f(x)" : getValue(x),
                        "f(x)'" : derivf,
                        "f_'(x)" : getFirstDevivativeOH(x, h),
                        "O(h)" : math.fabs(getFirstDevivativeOH(x, h) - derivf),
                        "f='(x)" :  getFirstDerivativeOH2(x, h),
                        "O(h^2)" : math.fabs( getFirstDerivativeOH2(x, h) - derivf),
                        "f(x)''" : secondderivf,
                        "f_(x)''" : getSecondDerivativeOH2(x, h),
                        "O(h^2)" : math.fabs(getSecondDerivativeOH2(x, h) - secondderivf)
                       }, ignore_index=True)
        
    return df

In [78]:
df_1 = f_1(df_1)

In [79]:
df_1.head(10)

x      f(x)      f(x)'   f_'(x)      O(h)    f='(x)    O(h^2)     f(x)''  \
0  0.0  1.000000   2.000000   1.8127  0.187300   1.96890  0.013000   4.000000   
1  0.1  1.221403   2.442806   2.2140  0.228806   2.40480  0.016389   4.885611   
2  0.2  1.491825   2.983649   2.7042  0.279449   2.93740  0.020701   5.967299   
3  0.3  1.822119   3.644238   3.3030  0.341238   3.58760  0.023525   7.288475   
4  0.4  2.225541   4.451082   4.0342  0.416882   4.38190  0.029836   8.902164   
5  0.5  2.718282   5.436564   4.9274  0.509164   5.35220  0.036873  10.873127   
6  0.6  3.320117   6.640234   6.0184  0.621834   6.53705  0.043532  13.280468   
7  0.7  4.055200   8.110400   7.3508  0.759600   7.98435  0.054200  16.220800   
8  0.8  4.953032   9.906065   8.9783  0.927765   9.75225  0.066870  19.812130   
9  0.9  6.049647  12.099295  10.9662  1.133095  11.91140  0.080410  24.198590   

   f_(x)''    O(h^2)  
0    4.013  0.013000  
1    4.902  0.016389  
2    5.988  0.020701  
3    7.312  0.023525  
4    8.932  0.029836  
5   10.910  0.036873  
6   13.324  0.043532  
7   16.275  0.054200  
8   19.879  0.066870  
9   24.279  0.080410

### 2) Пользуясь формулой $(6)$ в точке $x = 1$ вычислить разностную производную второго порядка
### аппроксимации функции $e^{2x}$, последовательно уменьшая шаг $h$ (например,вдвое) до тех пор, пока 
### фактическая погрешность не начнет возрастать. Определить $h$ оптимальное экпериментально и 
### теоретически, объяснить полученные результаты.Значения функции округлять до 5-ого знака после
### запятой, т. е. $ε = 5 · 10^{−6}$



In [80]:
df_info = pd.DataFrame(columns = ["h", 
                               "f",
                               "погрешность"]) 

In [81]:
# вывод поиска оптимального h для первой производной с первым порядком точности
def f_2(df, n):
    x = 1.0
    h = 0.2
    for i in range(n):
        h = h / 2
        df = df.append({"h" : h, 
                        "f" : getFirstDerivativeOH2(x, h),
                        "погрешность" : getExpFirstDerivative(x) - getFirstDerivativeOH2(x, h)
                       }, ignore_index=True)
    return df    

In [82]:
df_2 = f_2(df_info, 6) 
df_3 = f_2(df_info, 5)
df_4 = f_2(df_info, 4)

In [83]:
df_2

h        f  погрешность
0  0.100000  14.5484     0.229712
1  0.050000  14.7249     0.053212
2  0.025000  14.7650     0.013112
3  0.012500  14.7744     0.003712
4  0.006250  14.7768     0.001312
5  0.003125  14.7744     0.003712

In [84]:
df_3

h        f  погрешность
0  0.10000  14.5484     0.229712
1  0.05000  14.7249     0.053212
2  0.02500  14.7650     0.013112
3  0.01250  14.7744     0.003712
4  0.00625  14.7768     0.001312

In [85]:
df_4

h        f  погрешность
0  0.1000  14.5484     0.229712
1  0.0500  14.7249     0.053212
2  0.0250  14.7650     0.013112
3  0.0125  14.7744     0.003712

## Как видим пркатически лучший $h = 0.00625$

### 3) Предполагается заданной таблица значений функции в равноотстоящих узлах $x_i$, $i =0, . . . , n.$
### Требуется дифференцированием интерполяционного многочлена в форме Ньютона
### получить формулу численного дифференцирования для вычисления приближенного
### значения первой производной с третьим порядком аппроксимации в точке $x = x_n$.
###  Применить формулу для вычисления производной, сравнить с точным значением.

<img src="3(1).jpg">
<img src="3(2).jpg">

In [86]:
df_task_3 = pd.DataFrame(columns = ["h", 
                               "f",
                               "погрешность"]) 

In [87]:
def f_3(df, n):
    x = 1.0
    h = 0.2
    for i in range(n):
        h = h / 2
        df = df.append({"h" : h, 
                        "f" : getFirstDerivativeOH3(x, h),
                        "погрешность" : getExpFirstDerivative(x) - getFirstDerivativeOH3(x, h)
                       }, ignore_index=True)
    return df  

In [88]:
df_task_3 = f_3(df_task_3, 15) 

In [89]:
df_task_3

h          f  погрешность
0   0.100000   8.131017     6.647096
1   0.050000  10.951200     3.826912
2   0.025000  12.719400     2.058712
3   0.012500  13.710800     1.067312
4   0.006250  14.233333     0.544779
5   0.003125  14.506667     0.271446
6   0.001563  14.642133     0.135979
7   0.000781  14.707200     0.070912
8   0.000391  14.724267     0.053846
9   0.000195  14.856533    -0.078421
10  0.000098  14.779733    -0.001621
11  0.000049  14.813867    -0.035754
12  0.000024  14.267733     0.510379
13  0.000012  16.247467    -1.469354
14  0.000006  14.745600     0.032512